### PR TITLE
Add SupportedOSTypes function to Components

### PR DESCRIPTION
**Pull request by Brian-McM**
This forces every component to state what operating systems are supported for the objects being rendered. If a single OS is returned, then a node selector is used (to avoid any unnecessary disruption of the pods since node selectors were previously used), but if there are multiple OS types, then node affinity is used. The label for the selector / affinity is the "kuberentes.io/os" label.

## For PR author

- [x] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [x] Milestone set according to targeted release.
- [x] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and delete the branch after I've merged it.

**Review by tmjd — APPROVED** [review]
LGTM

**Comment by marvin-tigera** [comment]
I'm sorry but I failed delete the branch after merging the pull request.

**Pull request merged**
